src/compton_tools/cskrw.cc

```diff
@@ -0,0 +1,1023 @@
+  std::vector<FP> derivs;
+};
+
+Sparse_Compton_Data::Sparse_Compton_Data() : first_groups(), indexes(), data(), derivs() {}
+
+struct Dense_Compton_Data {
+  UINT numEvals;
+  UINT numTs;
+  UINT numGroups;
+  UINT numLegMoments;
+  std::vector<FP> groupBdrs;
+  std::vector<FP> Ts;
+  // [eval, moment, T, gfrom, gto]
+  std::vector<FP> data;
+  // [eval, moment, T, gfrom, gto]
+  std::vector<FP> derivs;
+
+  void resize(UINT numfiles, std::string filename);
+  void read_from_file(UINT eval, std::string filename, bool isnonlin);
+  void compute_nonlinear_difference();
+  void compute_temperature_derivatives();
+  void write_sparse_binary(std::string fileout);
+  void print_contents(int verbosity, int precision);
+
+private:
+  Sparse_Compton_Data copy_to_sparse();
+  void print_sparse(const Sparse_Compton_Data &sd);
+  void write_binary(std::string fileout, Sparse_Compton_Data &sd);
+};
+
+// resize data and set sizes
+// numfiles is number of Compton files; filename is one such file
+void Dense_Compton_Data::resize(UINT numfiles, std::string filename) {
+  // Reserve space for nl difference
+  UINT numderived = (numfiles >= 4) ? 1 : 0;
+  numEvals = numfiles + numderived;
+
+  // Read first line of filename to get sizes
+  std::ifstream f(filename);
+  Insist(f.is_open(), "Unable to open " + filename);
+
+  // Line 1: sizes
+  UINT numTbreakpoints = 0; // not used
+  numTs = 0;
+  numGroups = 0;
+  numLegMoments = 0;
+  f >> numTbreakpoints >> numTs >> numGroups >> numLegMoments;
+
+  // Set vector lengths
+  groupBdrs.resize(numGroups + 1, 0.0);
+  Ts.resize(numTs, 0.0);
+  UINT sz = numEvals * numLegMoments * numTs * numGroups * numGroups;
+  data.resize(sz, 0.0);
+  derivs.resize(sz, 0.0);
+
+  f.close();
+}
+
+// read the entire contents of one file
+void Dense_Compton_Data::read_from_file(UINT eval, std::string filename, bool isnonlin) {
+  Insist(eval < numEvals, "eval must be < numEvals");
+  std::ifstream f(filename);
+  Insist(f.is_open(), "Unable to open " + filename);
+
+  // Normalization / unit change
+  const FP mtocm = 100.0;
+  const FP classical_electron_radius = mtocm * rtt_units::classicalElectronRadiusSI; // cm
+
+  // Normalization constants for raw CSK data:
+  // CSK to cross section: 2 * pi * classicalElectronRadius^2 / 4
+  // cross section to opacity: Zbar_over_A * avogadrosNumber
+  // opacity to micro xs: 1/avogadrosNumber
+  //
+  // convert from CSK data to opacity (cm^2/mole)
+  const FP csk_opac_norm = 0.25 * 2 * rtt_units::PI * classical_electron_radius *
+                           classical_electron_radius * rtt_units::AVOGADRO;
+  Ensure(soft_equiv(csk_opac_norm, 2 * 0.037558, 1e-4));
+  //
+  // convert from opacity (cm^2/mole) to micro xs (cm^2)
+  //const FP csk_xs_conv = 1. / rtt_units::AVOGADRO;
+  //Ensure(soft_equiv(2 * 0.2003102 * csk_xs_conv, 6.6524587e-25, 1e-4));
+
+  // VALUES USED IN CSK'S PHYSICAL_CONSTANTS.HH
+  // Normalization for energy and nonlinear terms
+  const FP mec2 = 510.998;
```

Comment:
* Probably better to pull these values directly from csk.
* This file looks a bit rough, maybe just add some history and discussion to the top of the file.  It doesn't appear to be something that would be consumed by users.

Reply:
I'd like to have a code that in no way requires the CSK library. Added history and intent comments.